components/membership/membership-service-impl/src/test/kotlin/net/corda/membership/service/impl/CommandsRetryManagerTest.kt

```diff
@@ -0,0 +1,218 @@
+    }
+    private val messagingConfig = mock<SmartConfig>()
+    private val publisher = mock<Publisher>()
+    private val publisherFactory = mock<PublisherFactory> {
+        on { createPublisher(any(), same(messagingConfig)) } doReturn publisher
+    }
+    private val nowInMillis = 10000L
+    private val now = mock<Instant> {
+        on { toEpochMilli() } doReturn nowInMillis
+    }
+    private val clock = mock<Clock> {
+        on { instant() } doReturn now
+    }
+
+    private val manager = CommandsRetryManager(
+        coordinatorFactory,
+        publisherFactory,
+        messagingConfig,
+        clock,
+    )
+
+    @Test
+    fun `constructor will start the publisher`() {
+        verify(publisher).start()
+    }
+
+    @Test
+    fun `constructor will start the coordinator`() {
+        verify(coordinator).start()
+    }
+
+    @Test
+    fun `constructor will set the coordinator status to UP`() {
+        verify(coordinator).updateStatus(LifecycleStatus.UP)
+    }
+
+    @Test
+    fun `close will close the publisher`() {
+        manager.close()
+
+        verify(publisher).close()
+    }
+
+    @Test
+    fun `close will close the coordinator`() {
+        manager.close()
+
+        verify(coordinator).close()
+    }
+
+    @Test
+    fun `onPartitionSynced will add a timer for each command`() {
+        val states = (1..3).associate {
+            "key-$it" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            }
+        }
+
+        manager.onPartitionSynced(states)
+
+        verify(coordinator).setTimer(eq("retry-key-1"), eq(2000), any())
+        verify(coordinator).setTimer(eq("retry-key-2"), eq(2000), any())
+        verify(coordinator).setTimer(eq("retry-key-3"), eq(2000), any())
+    }
+
+    @Test
+    fun `onPartitionLost will cancel the timer for each command`() {
+        val states = (1..3).associate {
+            "key-$it" to mock<MembershipAsyncRequestState>()
+        }
+
+        manager.onPartitionLost(states)
+
+        verify(coordinator).cancelTimer("retry-key-1")
+        verify(coordinator).cancelTimer("retry-key-2")
+        verify(coordinator).cancelTimer("retry-key-3")
+    }
+
+    @Test
+    fun `onPostCommit will cancel the timers for removed commands`() {
+        val states = mapOf(
+            "key-1" to null,
+            "key-2" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            },
+            "key-3" to null,
+            "key-4" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            },
+        )
+
+        manager.onPostCommit(states)
+
+        verify(coordinator).cancelTimer("retry-key-1")
+        verify(coordinator).cancelTimer("retry-key-3")
```

Comment:
Could be good to check the timer isn't called for non-null keys
```suggestion
        verify(coordinator).cancelTimer("retry-key-1")
        verify(coordinator, never()).cancelTimer("retry-key-2")
        verify(coordinator).cancelTimer("retry-key-3")
        verify(coordinator, never()).cancelTimer("retry-key-4")
```